Ed-Fi-Exchange-OSS/EdFi-Project-Buzz
# [FIF-300] Unified installer for Buzz (Partial push to unblock team)

## OVERVIEW

FIF-300 currently blocks other dev ops related tickets. FIF-300 in its current state provides the unified installer for EdFi.Buzz in EdFi.Buzz.Installer and unblocks the tasks that depend on the PowerShell script for building the package, and executing the Octopus Deploy script.

Therefore, we are submitting FIF-300 currently to unblock FIF-291 and FIF-299. Known issues are listed below.

## TO TEST

- Create a `C:\nugetRepo folder` as your local NuGet repository. This is the stand-in for FIF_291's Azure Artifact Repository. We can change this location in the configuration.json of the EdFi.Buzz.Installer. configuration.json: artifactRepo
- Update the EdFi.Buzz.Installer\eng\Windows\configuration.json with any customizations for your environment or testing. 
- Run `build-installer.ps1` for each of the Buzz apps in their `\eng\` folder and copy the nupkg files to the `C:\nugetRepo` folder from step 1.
- In a PowerShell console, set location to the EdFi.Buzz.Installer\eng\Windows folder. NOTE: There are issues with PowerShell Core and the WebAdministration module, so use 5.1 or below.
- To ensure that you are downloading the NuGet packages each time, clear out any existing packages in the c:\temp\packages. You might also need to clear local caches - `nuget locals all -clear`. (update: for clarity)
- Execute `.\install.ps1` as administrator to install all applications. Confirm that you have 3 EdFi services - ETL, API and UI. NOTE: Our webs will not be executing yet. The Web and UI are running from the wrong location at present. They are being moved to `C:\inetpub\Ed-Fi\Buzz` in a later PR.
- Execute `.\uninstall.ps1` to uninstall all applications. All Ed-Fi Services should be stopped and removed from Services.
NOTE: The webs will NOT be in a running state. ODS tools is a dotnet stack and doesn't need NGINX. There is an additional step to move the AppCommon dependency to UI and API, and have those installs call the AppCommon IIS registration functions.

[FIF-300] Known issues remaining 

- API and UI installs are incomplete.

1. We need UI and API to have a NuGet dependency on the EdFi.Installer.AppCommon package to coordinate their installation into IIS.

    API and UI are currently Node JS webs running inside NGINX.
    We are using the EdFi.Installer.AppCommon which installs webs into IIS after moving the files from a target directory to their inetpub directory.
    For NGINX to work, we will need the Application Request Routing Reverse proxy to point our IIS Web sites to our ports.
    Additionally, our Buzz app eng\install.ps1 PowerShell scripts should initiate the install from EdFi.Installer.AppCommon to coordinate their own installation.
    Presently, we install our services and then the unified Buzz installer (EdFi.Buzz.Installer) calls AppCommon after
